Ensure file is closed promptly in case of error

rxxg opened the pull request:
Fixes #35566.  Replaces #35567.

rxxg committed:
Ensure resource closure in all exceptional circumstances during construction (#35566)

rxxg commented:
Not quite sure how the test failure is going to work in this case:  previous issues (e.g. #13932) mention ResourceWarnings but I don't have time to set up a full local build environment to duplicate the tests.

WillAyd commented on pandas/tests/io/sas/test_sas7bdat.py:
```diff
@@ -217,6 +217,14 @@ def test_zero_variables(datapath):
+def test_corrupt_read(datapath):
+    # We don't really care about the exact failure, the important thing is
+    # that the resource should be cleaned up afterwards (BUG #35566)
+    fname = datapath("io", "sas", "data", "corrupt.sas7bdat")
```
Can you test that the file is actually closed?

rxxg replied:
If you have any ideas how to test for that I am all ears!
Certain builds will emit a ResourceWarning but it is only for information - I have visually monitored this, and this patch adds a new warning with the test case and removes it with the change to sasreader.py
(It looks like ResourceWarnings are periodically manually cleaned up, eg #13292, #23680, ..., but there is no mechanism in place for systematically catching them.)

jreback replied:
yeah this is fine; we don't really deal well with ResourceWarnings; happy to have a patch which elevates these to fail the build (separate PR)

simonjayhawkins commented:
copied the milestone from #35567. feel free to amend as appropriate.

jreback reviewed (changes requested):
can you add a whatsnew note in 1.2, bug fix section for I/O

pep8speaks commented:
Hello @rxxg! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/io/sas/sas7bdat.py`](https://github.com/pandas-dev/pandas/blob/422076248e29ee04fbbdeacfbfa919db38050f6a/pandas/io/sas/sas7bdat.py):

> [Line 148:9](https://github.com/pandas-dev/pandas/blob/422076248e29ee04fbbdeacfbfa919db38050f6a/pandas/io/sas/sas7bdat.py#L148): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'

* In the file [`pandas/io/sas/sas_xport.py`](https://github.com/pandas-dev/pandas/blob/422076248e29ee04fbbdeacfbfa919db38050f6a/pandas/io/sas/sas_xport.py):

> [Line 272:9](https://github.com/pandas-dev/pandas/blob/422076248e29ee04fbbdeacfbfa919db38050f6a/pandas/io/sas/sas_xport.py#L272): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'

rxxg commented:
Added the What's New note.
I also added a try/except to ensure correctness under more circumstances, but it raised two PEP8 flags (which I silenced).  Happy to discuss my justification.

jreback commented on pandas/io/sas/sas_xport.py:
```diff
@@ -267,7 +267,11 @@ def __init__(
-        self._read_header()
+        try:
+            self._read_header()
+        except:  # noqa: E722  OK because exception re-raised
```
*never* use bare exceptions, use
```
try:
   ....
except Exception:
   # close on error reading header...
   ....
```

rxxg replied:
I can do that, but what about errors that don't inherit from Exception?
I note that the PEP says:

> A good rule of thumb is to limit use of bare 'except' clauses to two cases:
>    If the exception handler will be printing out or logging the traceback; at least the user will be aware that an error has occurred.
 >   If the code needs to do some cleanup work, but then lets the exception propagate upwards with raise. try...finally can be a better way to handle this case.

(try...finally obviously isn't applicable here since we only want to close on error)

jreback replied:
simply use Exception

rxxg replied:
OK, I note that exceptions that don't inherit from Exception are supposed to be system-exiting, so we can suppose some higher-level process will clean up for us in this case.

WillAyd commented:
/azp run

jbrockmendel commented:
I'm fine with this implementation, but curious: wouldnt the idiomatic way to do this be with a contextmanager?

jreback commented on pandas/io/sas/sasreader.py:
```diff
@@ -136,8 +136,8 @@ def read_sas(
-        reader.close()
-    return data
+    try:
+        return reader.read()
```
is this needed that you are catching error (and closing) at the lower level? maybe *only* catch here

rxxg replied:
The close at the lower level (sas7bdat.py, sas_xport.py) is to match with the get_filepath_or_buffer call 10 lines earlier, and also with various calls to close scatter around those files.
It is possible (if both 'lower level' file are only ever used via sasreader.py) that all close calls are unnecessary lower down, but I don't think they do any harm so I would prefer to leave the code symmetrical for the moment (open/close) and potentially get rid of all the close calls in a separate PR.
What do you think?

jreback replied:
ok fair enough

jreback reviewed (approved):
thanks @rxxg happy to take a followup to see if you can remove the lower level try/excepts
